File: ios/WE/Session.swift
import Foundation
import Supabase

@MainActor
final class Session: ObservableObject {
    enum Status { case loading, signedOut, needsCouple, ready }

    @Published var status: Status = .loading
    @Published var user: (id: String, name: String)?
    @Published var couple: (id: String, joinCode: String)?
    @Published var members: [Member] = []
    @Published var viewer: PersonID = ""
    @Published var records: [Record] = []
    @Published var reflections: [Reflection] = []
    @Published var busy = false
    @Published var errorText: String?
    @Published var info: String?

    let client = SupabaseClient(supabaseURL: Config.supabaseURL, supabaseKey: Config.supabaseAnonKey)

    var partner: Member? { members.first { $0.id != viewer } }
    func member(_ id: PersonID?) -> Member? { id.flatMap { mid in members.first { $0.id == mid } } }

    // MARK: - Auth

    func start() async {
        do {
            _ = try await client.auth.session
            await load()
        } catch {
            status = .signedOut
        }
    }

    func signIn(email: String, password: String) async {
        busy = true; errorText = nil; info = nil
        do {
            try await client.auth.signIn(email: email.trimmed, password: password)
            await load()
        } catch {
            errorText = friendly(error)
        }
        busy = false
    }

    func signUp(name: String, email: String, password: String) async {
        busy = true; errorText = nil; info = nil
        do {
            let res = try await client.auth.signUp(
                email: email.trimmed,
                password: password,
                data: ["name": .string(name.trimmed)]
            )
            if res.session == nil {
                info = "Check your email to confirm, then sign in."
            } else {
                await load()
            }
        } catch {
            errorText = friendly(error)
        }
        busy = false
    }

    func signOut() async {
        try? await client.auth.signOut()
        user = nil; couple = nil; members = []; records = []; reflections = []
        status = .signedOut
    }

    // MARK: - Pairing

    func createCouple() async {
        await run {
            let res: CoupleResult = try await self.client.rpc("create_couple").execute().value
            _ = res
            await self.load()
        }
    }

    func joinCouple(code: String) async {
        await run {
            _ = try await self.client.rpc("join_couple", params: JoinParam(p_code: code.trimmed)).execute()
            await self.load()
        }
    }

    // MARK: - Consent actions (server-enforced RPCs)

    func request(_ id: String) async { await rpc("request_reveal", InsightParam(p_insight: id)) }
    func accept(_ id: String) async { await rpc("accept_reveal", InsightParam(p_insight: id)) }
    func decline(_ id: String) async { await rpc("decline_reveal", InsightParam(p_insight: id)) }
    func withdraw(_ id: String) async { await rpc("withdraw_reveal", InsightParam(p_insight: id)) }
    func dismiss(_ id: String) async { await rpc("dismiss_suggestion", InsightParam(p_insight: id)) }

    func submit(_ id: String, choice: String, note: String?) async {
        await rpc("submit_response", SubmitParam(p_insight: id, p_choice: choice, p_note: note))
    }

    func settle(_ id: String, type: ResolutionType, choice: String?) async {
        await rpc("resolve_insight", ResolveParam(p_insight: id, p_type: type.rawValue, p_choice: choice))
    }

    private func rpc<P: Encodable>(_ fn: String, _ params: P) async {
        await run {
            _ = try await self.client.rpc(fn, params: params).execute()
            await self.load()
        }
    }

    private func run(_ work: @escaping () async throws -> Void) async {
        errorText = nil
        do { try await work() } catch { errorText = friendly(error) }
    }

    // MARK: - Load everything for the couple, then build projected records

    func load() async {
        do {
            let uid = try await client.auth.session.user.id.uuidString.lowercased()

            let profile: [ProfileRow] = try await client.from("profiles")
                .select("id,name").eq("id", value: uid).execute().value
            user = (uid, profile.first?.name ?? "You")
            viewer = uid

            let mine: [MembershipRow] = try await client.from("couple_members")
                .select("couple_id").eq("profile_id", value: uid).execute().value
            guard let coupleId = mine.first?.couple_id else {
                status = .needsCouple
                members = []; records = []; reflections = []
                return
            }

            async let coupleRows: [CoupleRow] = client.from("couples")
                .select("id,join_code").eq("id", value: coupleId).execute().value
            async let memberRows: [MemberRow] = client.from("couple_members")
                .select("profile_id,hue,profiles(name)").eq("couple_id", value: coupleId).execute().value
            async let insightRows: [InsightRow] = client.from("insights")
                .select("*").eq("couple_id", value: coupleId).order("sort").execute().value
            async let consentRows: [ConsentRow] = client.from("insight_consent").select("*").execute().value
            async let responseRows: [ResponseRow] = client.from("responses").select("*").execute().value
            async let dismissalRows: [DismissalRow] = client.from("dismissals").select("*").execute().value
            async let reflectionRows: [ReflectionRow] = client.from("reflections")
                .select("*").eq("owner_id", value: uid).execute().value

            let cRows = try await coupleRows
            let mRows = try await memberRows
            let iRows = try await insightRows
            let conRows = try await consentRows
            let rRows = try await responseRows
            let dRows = try await dismissalRows
            let reflRows = try await reflectionRows

            couple = cRows.first.map { ($0.id, $0.join_code) }
            members = mRows.map { Member(id: $0.profile_id, name: $0.profiles?.name ?? "Partner", hue: Hue(rawValue: $0.hue) ?? .burgundy) }
            let partnerID = members.first { $0.id != uid }?.id

            let consentByInsight = Dictionary(uniqueKeysWithValues: conRows.map { ($0.insight_id, $0) })
            var responsesByInsight: [String: [ResponseRow]] = [:]
            for r in rRows { responsesByInsight[r.insight_id, default: []].append(r) }
            let dismissed = Set(dRows.map { $0.insight_id })

            records = iRows.map { i in
                let c = consentByInsight[i.id]
                let rows = responsesByInsight[i.id] ?? []
                func resp(_ pid: String?) -> PersonResponse {
                    guard let pid, let row = rows.first(where: { $0.profile_id == pid }) else { return PersonResponse() }
                    return PersonResponse(status: ResponseStatus(rawValue: row.status) ?? .none, choice: row.choice, note: row.note)
                }
                var responses: [PersonID: PersonResponse] = [uid: resp(uid)]
                if let partnerID { responses[partnerID] = resp(partnerID) }

                let state = InsightState(
                    visibility: c.flatMap { Visibility(rawValue: $0.visibility) } ?? .shared,
                    owner: c?.owner_id,
                    readiness: c.flatMap { Readiness(rawValue: $0.readiness) } ?? .idle,
                    initiator: c?.initiator_id,
                    requestedAt: c?.requested_at.flatMap(parseTimestamp),
                    acceptedAt: c?.accepted_at.flatMap(parseTimestamp),
                    responses: responses,
                    dismissedBy: dismissed.contains(i.id) ? [uid] : [],
                    resolution: c?.resolution_type.flatMap { ResolutionType(rawValue: $0) }.map { Resolution(type: $0, choice: c?.resolution_choice) }
                )
                let insight = Insight(id: i.id, kind: InsightKind(rawValue: i.kind) ?? .logistical,
                                      domain: Domain(rawValue: i.domain) ?? .life, present: i.present,
                                      title: i.title, body: i.body, evidence: i.evidence, source: i.source, options: i.options)
                return Record(insight: insight, state: state)
            }

            reflections = reflRows.map {
                Reflection(id: $0.id, owner: $0.owner_id, domain: Domain(rawValue: $0.domain) ?? .life,
                           kind: ReflectionKind(rawValue: $0.kind) ?? .reflection, text: $0.text)
            }

            status = .ready
        } catch {
            errorText = friendly(error)
            if case .loading = status { status = .signedOut }
        }
    }
}

// MARK: - RPC params

private struct InsightParam: Encodable { let p_insight: String }
private struct JoinParam: Encodable { let p_code: String }
private struct SubmitParam: Encodable { let p_insight: String; let p_choice: String; let p_note: String? }
private struct ResolveParam: Encodable { let p_insight: String; let p_type: String; let p_choice: String? }
private struct CoupleResult: Decodable { let couple_id: String; let join_code: String }

// MARK: - DB rows (snake_case, decoded directly from PostgREST)

private struct ProfileRow: Decodable { let id: String; let name: String }
private struct MembershipRow: Decodable { let couple_id: String }
private struct CoupleRow: Decodable { let id: String; let join_code: String }
private struct NameOnly: Decodable { let name: String }
private struct MemberRow: Decodable { let profile_id: String; let hue: String; let profiles: NameOnly? }
private struct InsightRow: Decodable {
    let id: String; let kind: String; let domain: String; let present: Bool
    let title: String; let body: String; let evidence: String; let source: String; let options: [String]
}
private struct ConsentRow: Decodable {
    let insight_id: String; let visibility: String; let owner_id: String?
    let readiness: String; let initiator_id: String?
    let requested_at: String?; let accepted_at: String?
    let resolution_type: String?; let resolution_choice: String?; let resolved_at: String?
}
private struct ResponseRow: Decodable { let insight_id: String; let profile_id: String; let status: String; let choice: String?; let note: String? }
private struct DismissalRow: Decodable { let insight_id: String; let profile_id: String }
private struct ReflectionRow: Decodable { let id: String; let owner_id: String; let domain: String; let kind: String; let text: String }

// MARK: - Helpers

private let isoFormatter: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return f
}()

private func parseTimestamp(_ s: String) -> Date? {
    isoFormatter.date(from: s) ?? ISO8601DateFormatter().date(from: s)
}

private func friendly(_ error: Error) -> String {
    let msg = (error as NSError).localizedDescription
    return msg.isEmpty ? "Something went wrong." : msg
}

private extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

File: lib/session.ts
"use client";

import { create } from "zustand";
import { supabase, supabaseConfigured } from "./supabase";
import type { InsightRecord, InsightState, PersonResponse, Reflection } from "./types";

export type Member = { id: string; name: string; hue: "burgundy" | "sage" };
export type Status = "loading" | "unconfigured" | "signedOut" | "needsCouple" | "ready";

interface DbConsent {
  insight_id: string;
  visibility: "private" | "shared" | "mutual";
  owner_id: string | null;
  readiness: "idle" | "requested" | "accepted" | "declined" | "withdrawn";
  initiator_id: string | null;
  requested_at: string | null;
  accepted_at: string | null;
  resolution_type: "settled" | "released" | "leftOpen" | null;
  resolution_choice: string | null;
}
interface DbResponse {
  insight_id: string;
  profile_id: string;
  status: "none" | "draft" | "submitted" | "revealed";
  choice: string | null;
  note: string | null;
}

interface SessionState {
  status: Status;
  user: { id: string; name: string } | null;
  couple: { id: string; joinCode: string } | null;
  members: Member[];
  viewer: string;
  records: InsightRecord[];
  reflections: Reflection[];
  busy: boolean;
  error: string | null;
  info: string | null;
  announcement: string;

  init: () => Promise<void>;
  signIn: (email: string, password: string) => Promise<void>;
  signUp: (email: string, password: string, name: string) => Promise<void>;
  signOut: () => Promise<void>;
  createCouple: () => Promise<void>;
  joinCouple: (code: string) => Promise<void>;
  load: () => Promise<void>;

  request: (id: string) => Promise<void>;
  accept: (id: string) => Promise<void>;
  decline: (id: string) => Promise<void>;
  withdraw: (id: string) => Promise<void>;
  submit: (id: string, choice: string, note?: string) => Promise<void>;
  settle: (id: string, type: "settled" | "released" | "leftOpen", choice?: string) => Promise<void>;
  dismiss: (id: string) => Promise<void>;

  memberById: (id?: string | null) => Member | undefined;
  partnerId: () => string | undefined;
}

let realtimeBound = false;

export const useSession = create<SessionState>((set, get) => ({
  status: "loading",
  user: null,
  couple: null,
  members: [],
  viewer: "",
  records: [],
  reflections: [],
  busy: false,
  error: null,
  info: null,
  announcement: "",

  memberById: (id) => (id ? get().members.find((m) => m.id === id) : undefined),
  partnerId: () => {
    const { members, viewer } = get();
    return members.find((m) => m.id !== viewer)?.id;
  },

  init: async () => {
    if (!supabaseConfigured) {
      set({ status: "unconfigured" });
      return;
    }
    const { data } = await supabase.auth.getSession();
    if (!data.session) {
      set({ status: "signedOut" });
    } else {
      await get().load();
    }
    supabase.auth.onAuthStateChange((_event, session) => {
      if (!session) {
        set({ status: "signedOut", user: null, couple: null, members: [], records: [], reflections: [] });
      }
    });
  },

  signIn: async (email, password) => {
    set({ busy: true, error: null, info: null });
    const { error } = await supabase.auth.signInWithPassword({ email: email.trim(), password });
    if (error) return set({ busy: false, error: error.message });
    await get().load();
    set({ busy: false });
  },

  signUp: async (email, password, name) => {
    set({ busy: true, error: null, info: null });
    const { data, error } = await supabase.auth.signUp({
      email: email.trim(),
      password,
      options: { data: { name: name.trim() } },
    });
    if (error) return set({ busy: false, error: error.message });
    if (!data.session) {
      // Email confirmation is on for this project — no session yet.
      return set({
        busy: false,
        info: "Check your email to confirm, then sign in.",
      });
    }
    await get().load();
    set({ busy: false });
  },

  signOut: async () => {
    await supabase.auth.signOut();
    set({ status: "signedOut", user: null, couple: null, members: [], records: [], reflections: [] });
  },

  createCouple: async () => {
    set({ busy: true, error: null });
    const { error } = await supabase.rpc("create_couple");
    if (error) return set({ busy: false, error: error.message });
    await get().load();
    set({ busy: false, announcement: "Your shared space is ready." });
  },

  joinCouple: async (code) => {
    set({ busy: true, error: null });
    const { error } = await supabase.rpc("join_couple", { p_code: code.trim() });
    if (error) return set({ busy: false, error: error.message });
    await get().load();
    set({ busy: false, announcement: "You're in. This is yours together now." });
  },

  load: async () => {
    const { data: auth } = await supabase.auth.getUser();
    const u = auth.user;
    if (!u) return set({ status: "signedOut" });

    const { data: profile } = await supabase.from("profiles").select("id,name").eq("id", u.id).maybeSingle();
    const user = { id: u.id, name: profile?.name ?? "You" };

    const { data: myMembership } = await supabase
      .from("couple_members")
      .select("couple_id")
      .eq("profile_id", u.id)
      .maybeSingle();

    if (!myMembership) {
      return set({ status: "needsCouple", user, viewer: u.id, couple: null, members: [], records: [], reflections: [] });
    }

    const coupleId = myMembership.couple_id as string;

    const [coupleRes, membersRes, insightsRes, consentRes, responsesRes, dismissalsRes, reflectionsRes] =
      await Promise.all([
        supabase.from("couples").select("id,join_code").eq("id", coupleId).maybeSingle(),
        supabase.from("couple_members").select("profile_id,hue,profiles(name)").eq("couple_id", coupleId),
        supabase.from("insights").select("*").eq("couple_id", coupleId).order("sort"),
        supabase.from("insight_consent").select("*"),
        supabase.from("responses").select("*"),
        supabase.from("dismissals").select("*"),
        supabase.from("reflections").select("*").eq("owner_id", u.id),
      ]);

    const members: Member[] = (membersRes.data ?? []).map((m) => {
      // profiles may come back as an object or single-element array depending on the join.
      const p = m.profiles as unknown as { name?: string } | { name?: string }[] | null;
      const name = Array.isArray(p) ? p[0]?.name : p?.name;
      return { id: m.profile_id as string, hue: m.hue as Member["hue"], name: name ?? "Partner" };
    });
    const partnerId = members.find((m) => m.id !== u.id)?.id;

    const consentByInsight = new Map<string, DbConsent>();
    (consentRes.data as DbConsent[] | null)?.forEach((c) => consentByInsight.set(c.insight_id, c));

    const responsesByInsight = new Map<string, DbResponse[]>();
    (responsesRes.data as DbResponse[] | null)?.forEach((r) => {
      const arr = responsesByInsight.get(r.insight_id) ?? [];
      arr.push(r);
      responsesByInsight.set(r.insight_id, arr);
    });

    const dismissedSet = new Set<string>();
    (dismissalsRes.data as { insight_id: string }[] | null)?.forEach((d) => dismissedSet.add(d.insight_id));

    const records: InsightRecord[] = (insightsRes.data ?? []).map((i) => {
      const c = consentByInsight.get(i.id);
      const rows = responsesByInsight.get(i.id) ?? [];
      const mkResp = (pid?: string): PersonResponse => {
        const row = rows.find((r) => r.profile_id === pid);
        return row
          ? { status: row.status, choice: row.choice ?? undefined, note: row.note ?? undefined }
          : { status: "none" };
      };
      const responses: Record<string, PersonResponse> = { [u.id]: mkResp(u.id) };
      if (partnerId) responses[partnerId] = mkResp(partnerId);

      const state: InsightState = {
        visibility: c?.visibility ?? "shared",
        owner: c?.owner_id ?? undefined,
        readiness: c?.readiness ?? "idle",
        initiator: c?.initiator_id ?? undefined,
        requestedAt: c?.requested_at ? Date.parse(c.requested_at) : undefined,
        acceptedAt: c?.accepted_at ? Date.parse(c.accepted_at) : undefined,
        responses,
        dismissedBy: dismissedSet.has(i.id) ? [u.id] : [],
        resolution: c?.resolution_type
          ? { type: c.resolution_type, choice: c.resolution_choice ?? undefined, at: 0 }
          : undefined,
      };

      return {
        insight: {
          id: i.id,
          kind: i.kind,
          domain: i.domain,
          present: i.present,
          title: i.title,
          body: i.body,
          evidence: i.evidence,
          source: i.source,
          options: i.options,
        },
        state,
      };
    });

    const reflections: Reflection[] = (reflectionsRes.data ?? []).map((r) => ({
      id: r.id,
      owner: r.owner_id,
      domain: r.domain,
      kind: r.kind,
      text: r.text,
    }));

    set({
      status: "ready",
      user,
      viewer: u.id,
      couple: coupleRes.data ? { id: coupleRes.data.id, joinCode: coupleRes.data.join_code } : null,
      members,
      records,
      reflections,
    });

    // Bind realtime once: any change in the couple's shared tables triggers a reload.
    if (!realtimeBound) {
      realtimeBound = true;
      supabase
        .channel("we-couple")
        .on("postgres_changes", { event: "*", schema: "public", table: "insight_consent" }, () => get().load())
        .on("postgres_changes", { event: "*", schema: "public", table: "responses" }, () => get().load())
        .on("postgres_changes", { event: "*", schema: "public", table: "couple_members" }, () => get().load())
        .subscribe();
    }
  },

  request: async (id) => runRpc(set, get, "request_reveal", { p_insight: id }, "Brought to your partner. Waiting, without pressure."),
  accept: async (id) => runRpc(set, get, "accept_reveal", { p_insight: id }, "Opened together. This now lives between you."),
  decline: async (id) => runRpc(set, get, "decline_reveal", { p_insight: id }, "Held for later. Your partner won't be notified."),
  withdraw: async (id) => runRpc(set, get, "withdraw_reveal", { p_insight: id }, "Withdrawn. No record remains."),
  submit: async (id, choice, note) =>
    runRpc(set, get, "submit_response", { p_insight: id, p_choice: choice, p_note: note ?? null }, "Your answer is in."),
  settle: async (id, type, choice) =>
    runRpc(set, get, "resolve_insight", { p_insight: id, p_type: type, p_choice: choice ?? null },
      type === "leftOpen" ? "Left open. Nothing has been decided." : "Settled together."),
  dismiss: async (id) => runRpc(set, get, "dismiss_suggestion", { p_insight: id }, "Set aside, just for you."),
}));

async function runRpc(
  set: (partial: Partial<SessionState>) => void,
  get: () => SessionState,
  fn: string,
  args: Record<string, unknown>,
  announce: string
) {
  set({ error: null });
  const { error } = await supabase.rpc(fn, args);
  if (error) {
    set({ error: error.message });
    return;
  }
  set({ announcement: announce });
  await get().load();
}
